board/safety.h

```diff
@@ -522,36 +523,46 @@ bool steer_torque_cmd_checks(int desired_torque, int steer_req, const SteeringLi
     violation = true;
   }
 
-  // certain safety modes set their steer request bit low for one frame at a
+  // certain safety modes set their steer request bit low for one or more frame at a
   // predefined max frequency to avoid steering faults in certain situations
   bool steer_req_mismatch = (steer_req == 0) && (desired_torque != 0);
   if (steer_req_mismatch) {
-    // no torque if request bit isn't high
+    // no torque if request bit isn't high and safety mode doesn't allow tolerance
     if (!limits.has_steer_req_tolerance) {
       violation = true;
 
     } else {
-      // disallow torque cut if not enough recent matching steer_req messages
-      if (valid_steer_req_count < limits.min_valid_request_frames) {
-        violation = true;
-      }
+      if (invalid_steer_req_count == 0) {
+        // disallow torque cut if not enough recent matching steer_req messages
+        if (valid_steer_req_count < limits.min_valid_request_frames) {
+          violation = true;
+        }
 
-      // or we've cut torque too recently in time
-      uint32_t ts_elapsed = get_ts_elapsed(ts, ts_steer_req_mismatch_last);
-      if (ts_elapsed < limits.min_valid_request_rt_interval) {
-        violation = true;
+        // or we've cut torque too recently in time
+        uint32_t ts_elapsed = get_ts_elapsed(ts, ts_steer_req_mismatch_last);
+        if (ts_elapsed < limits.min_valid_request_rt_interval) {
+          violation = true;
+        }
+      } else {
+        // or we're cutting more frames consecutively than allowed
+        if (invalid_steer_req_count >= limits.max_invalid_request_frames) {
+          violation = true;
+        }
       }
 
       valid_steer_req_count = 0;
       ts_steer_req_mismatch_last = ts;
+      invalid_steer_req_count = MIN(invalid_steer_req_count + 1, limits.max_invalid_request_frames);
     }
   } else {
     valid_steer_req_count = MIN(valid_steer_req_count + 1, limits.min_valid_request_frames);
+    invalid_steer_req_count = 0;
   }
```

Comment:
can you refactor this a bit to check `limits.has_steer_req_tolerance` earlier? I think it would simplify this a bit

Reply:
flipped around, to me it looks a bit messier, but let me know what you think

Reply:
I think it's easier to follow now